From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: aswin-mcw <aswin.kalies@multicorewareinc.com>
Date: Wed, 9 Sep 2026 13:14:25 +0530
Subject: [PATCH 047/109] Cygwin: gendef: restore support for
 arch-specific cygwin.din files

(cherry picked from commit a9916dad8dc6d87c54a2f869de27beae1c77f457)
---
 winsup/cygwin/Makefile.am        | 8 ++++++--
 winsup/cygwin/aarch64/cygwin.din | 6 ++++++
 winsup/cygwin/x86_64/cygwin.din  | 6 ++++++
 3 files changed, 18 insertions(+), 2 deletions(-)
 create mode 100644 winsup/cygwin/aarch64/cygwin.din
 create mode 100644 winsup/cygwin/x86_64/cygwin.din

diff --git a/winsup/cygwin/Makefile.am b/winsup/cygwin/Makefile.am
index b00b5b080..806a4d51a 100644
--- a/winsup/cygwin/Makefile.am
+++ b/winsup/cygwin/Makefile.am
@@ -633,8 +633,12 @@ $(NEW_DLL_NAME): $(LDSCRIPT) libdll.a $(VERSION_OFILES) $(LIBSERVER)\
 # msys-2.0 import library
 toolopts=--cpu=@target_cpu@ --ar=@AR@ --as=@AS@ --nm=@NM@ --objcopy=@OBJCOPY@
 
-$(DEF_FILE): scripts/gendef cygwin.din
-	$(AM_V_GEN)$(srcdir)/scripts/gendef --cpu=@target_cpu@ --output-def=$(DEF_FILE) $(srcdir)/cygwin.din
+# Architecture-specific .din files
+ARCH_DIN = $(srcdir)/@target_cpu@/cygwin.din
+
+$(DEF_FILE): scripts/gendef cygwin.din $(wildcard $(ARCH_DIN))
+	$(AM_V_GEN)$(srcdir)/scripts/gendef --cpu=@target_cpu@ --output-def=$(DEF_FILE) \
+	  $(wildcard $(ARCH_DIN)) $(srcdir)/cygwin.din
 
 sigfe.s: $(DEF_FILE) tlsoffsets
 	@[ -s $@ ] || \
diff --git a/winsup/cygwin/aarch64/cygwin.din b/winsup/cygwin/aarch64/cygwin.din
new file mode 100644
index 000000000..194873e86
--- /dev/null
+++ b/winsup/cygwin/aarch64/cygwin.din
@@ -0,0 +1,6 @@
+# aarch64-specific exports
+# These symbols are available on AArch64 architecture
+
+LIBRARY "msys-2.0.dll" BASE=0x180040000
+
+EXPORTS
diff --git a/winsup/cygwin/x86_64/cygwin.din b/winsup/cygwin/x86_64/cygwin.din
new file mode 100644
index 000000000..12a49b009
--- /dev/null
+++ b/winsup/cygwin/x86_64/cygwin.din
@@ -0,0 +1,6 @@
+# x86_64-specific exports
+# These symbols are available on x64 architecture
+
+LIBRARY "cygwin1.dll" BASE=0x180040000
+
+EXPORTS
